File: coursework2/hw2.py
"""M3C 2018 Homework 2
Name: Tudor Mihai Trita Trita
CID: 01199397
TID: 167
Course: Mathematics G103 Year 3
"""

import numpy as np
import matplotlib.pyplot as plt
#from scipy import optimize
from m5 import nmodel as nm #assumes that hw2_dev.f90 has been
#compiled with: f2py -c hw2_dev.f90 -m m5
#from sklearn.neural_network import MLPClassifier
# May also use scipy, scikit-learn, and time modules as needed

def read_data(tsize=60000):
    """Read in image and label data from data.csv.
    The full image data is stored in a 784 x 70000 matrix, X
    and the corresponding labels are stored in a 70000 element array, y.
    The final 70000-tsize images and labels are stored in X_test
    and y_test, respectively.
    X,y,X_test, and y_test are all returned by the function.
    You are not required to use this function.
    """
    print("Reading data...") #may take 1-2 minutes
    Data=np.loadtxt('data.csv',delimiter=',')
    Data =Data.T
    X,y = Data[:-1,:]/255.,Data[-1,:].astype(int)%2 #rescale the image, convert the labels to 0s and 1s (For even and odd integers)
    Data = None

    # Extract testing data
    X_test = X[:,tsize:]
    y_test = y[tsize:]
    print("processed dataset")
    return X,y,X_test,y_test
#----------------------------

def snm_test(X,y,X_test,y_test,omethod,input=(None)):
    """Train single neuron model with input images and labels (i.e. use data in X and y), then compute and return testing error in test_error
    using X_test, y_test. The fitting parameters obtained via training should be returned in the 1-d array, fvec_f
    X: training image data, should be 784 x d with 1<=d<=60000
    y: training image labels, should contain d elements
    X_test,y_test: should be set as in read_data above
    omethod=1: use l-bfgs-b optimizer
    omethod=2: use stochastic gradient descent
    input: tuple, set if and as needed
    """
    n = X.shape[0]
    fvec = np.random.randn(n+1) #initial fitting parameters

    #Add code to train SNM and evaluate testing test_error
    d = X.shape[1]
    nm.nm_x = X
    nm.nm_y = y

    #Train snm using appropriate model
    if omethod==1:
        res = minimize(nm.snmodel,fvec,args=(d,),method='L-BFGS-B',jac=True)
        fvec_f = res.x
    elif omethod==2:
        fvec_f = nm.sgd(fvec,n,0,d,0.1)
    else:
        print("error, omethod must be 1 or 2")
        return None

    #Compute testing error
    z = np.dot(fvec_f[:-1],X_test) + fvec_f[-1]
    a_int = np.round(1.0/(1.0+np.exp(-z)))
    eps = np.abs(a_int - y_test)
    test_error = eps.sum()/y_test.size

    output = (None) #output tuple, modify as needed
    return fvec_f,test_error,output
#--------------------------------------------


def nnm_test(X,y,X_test,y_test,m,omethod,input=(None)):
    """Train neural network model with input images and labels (i.e. use data in X and y), then compute and return testing error (in test_error)
    using X_test, y_test. The fitting parameters obtained via training should be returned in the 1-d array, fvec_f
    X: training image data, should be 784 x d with 1<=d<=60000
    y: training image labels, should contain d elements
    X_test,y_test: should be set as in read_data above
    m: number of neurons in inner layer
    omethod=1: use l-bfgs-b optimizer
    omethod=2: use stochastic gradient descent
    input: tuple, set if and as needed
    """
    n = X.shape[0]
    fvec = np.random.randn(m*(n+2)+1) #initial fitting parameters

    #Add code to train NNM and evaluate testing error, test_error
    d = X.shape[1]
    nm.nm_x = X.copy()
    nm.nm_y = y.copy()

    #Train snm using appropriate model
    if omethod==1:
        res = minimize(nm.nnmodel,fvec,args=(n,m,d),method='L-BFGS-B',jac=True)
        fvec_f = res.x
    elif omethod==2:
        fvec_f = nm.sgd(fvec,n,m,d,0.1)
    else:
        print("error, omethod must be 1 or 2")
        return None

    nm.nm_x = X_test.copy()
    nm.nm_y = y_test.copy()

    test_error = nm.run_nnmodel(fvec_f,n,m,d)

    output = (None) #output tuple, modify as needed
    return fvec_f,test_error,output
#--------------------------------------------


def nm_analyze(mvalues,dvalues):
    """ Analyze performance of single neuron and neural network models
    on even/odd image classification problem
    Add input variables and modify return statement as needed.
    Should be called from
    name==main section below
    """

    #Read in data, and initialize arrays
    Xfull,yfull,X_test,y_test = read_data()
    dt = np.zeros((len(dvalues),len(mvalues),2))
    e_test = np.zeros((len(dvalues),len(mvalues),2))
    n = Xfull.shape[0]

    nm.nm_x_test = X_test
    nm.nm_y_test = y_test
    d_test = y_test.size

    #loop through d and m values storing wall time and testing error
    for i,d in enumerate(dvalues):
        nm.nm_x = Xfull[:,:d]
        nm.nm_y = yfull[:d]
        for j,m in enumerate(mvalues):
            for k,omethod in enumerate([1,2]):
                print("i,j,k=",i,j,k)
                args_net = (n,m,d)
                if m==0: #SNM
                    fvec = np.random.randn(n+1)
                    args = (d,)
                    if omethod==1: #BFGS
                        t1=time()
                        res = minimize(nm.snmodel,fvec,args,method='L-BFGS-B',jac=True)
                        fvec_f = res.x
                        t2=time()

                    else: #SGD
                        t1=time()
                        fvec_f = nm.sgd(fvec,n,m,d,0.1)
                        t2 = time()
                    #Compute testing error
                    z = np.dot(fvec_f[:-1],X_test) + fvec_f[-1]
                    a_int = np.round(1.0/(1.0+np.exp(-z)))
                    eps = np.abs(a_int - y_test)
                    e_test[i,j,k] = eps.sum()/y_test.size
                else: #NNM
                    fvec = np.random.randn(m*(n+2)+1)
                    args = (n,m,d)
                    if omethod==1: #BFGS
                        t1=time()
                        res = minimize(nm.nnmodel,fvec,args,method='L-BFGS-B',jac=True)
                        fvec_f = res.x
                        t2=time()

                    else: #SGD
                        t1=time()
                        fvec_f = nm.sgd(fvec,n,m,d,0.1)
                        t2 = time()

                    e_test[i,j,k] = nm.run_nnmodel(fvec_f,n,m,d_test)
                dt[i,j,k] = t2-t1


    plt.figure()
    plt.semilogx(dt,e_test[:,:,0],'x--')
    plt.xlabel('time (s)')
    plt.ylabel('testing error')
    plt.title('Testing error vs. wall time for l-bfgs-b classification calculations \n d=%s' %str(dvalues))
    plt.legend(('SNM','m=1','2','4'))

    return (e_test,dt)
#--------------------------------------------



if __name__ == '__main__':
    #The code here should call analyze and generate the
    #figures that you are submitting with your code

    #output = nm_analyze()

    #Getting data
    X,y,X_test,y_test = read_data()

    figure = [True,False]

    #ADD ANALYSE:
    nm_analyze(X,y,X_test,y_test,figure)

File: coursework2/hw2.f90
!M3C 2018 Homework 2
!Name: Tudor Mihai Trita Trita
!CID: 01199397
!TID: 167
!Course: Mathematics G103 Year 3

module nmodel
  implicit none
  real(kind=8), allocatable, dimension(:,:) :: nm_x
  integer, allocatable, dimension(:) :: nm_y

contains

subroutine data_init(n,d)
  implicit none
  integer, intent(in) :: n,d
  if (allocated(nm_x)) deallocate(nm_x)
  if (allocated(nm_y)) deallocate(nm_y)
  allocate(nm_x(n,d),nm_y(d))
end subroutine data_init


subroutine snmodel(fvec,n,d,c,cgrad)
  implicit none
  integer, intent(in) :: n,d !training data sizes
  real(kind=8), dimension(n+1), intent(in) :: fvec !fitting parameters
  real(kind=8), intent(out) :: c !cost
  real(kind=8), dimension(n+1), intent(out) :: cgrad !gradient of cost
  !Declare other variables as needed
  real(kind=8) :: dinv,dcdb
  real(kind=8),dimension(d) :: z,a,e,gamma,dadb,eg
  real(kind=8), dimension(n) :: dcdw

  dinv = 1.d0/dble(d)

  !Compute inner layer activation
  z = matmul(fvec(1:n),nm_x) + fvec(n+1)
  a = 1.d0/(1.d0+exp(-z))

  !Compute cost
  gamma = a*(1.d0-a)
  e = a-nm_y
  c = 0.5d0*dinv*sum(e**2)

  !Compute gradient of cost
  eg = e*gamma
  cgrad(n+1) = dinv*sum(eg) !dcdb
  cgrad(1:n) = dinv*matmul(nm_x,eg) !dcdw

end subroutine snmodel

!Following subroutine is used to find activation data for snmodel, to compute test_error
subroutine tst(fvec,X_test,y_test,n,d,terror)
  implicit none
  integer, intent(in) :: n, d
  real(kind=8), dimension(n+1), intent(in) :: fvec
  real(kind=8), dimension(n,d), intent(in) :: X_test
  integer, dimension(d), intent(in) :: y_test
  real(kind=8), intent(out) :: terror
  integer :: i1, ncct
  real(kind=8), dimension(d) :: z, activ
  ncct = 0
  do i1 = 1,d
    z(i1) = sum(fvec(1:n)*X_test(1:n,i1)) + fvec(n+1)
    activ(i1) = 1.d0/(1.d0 + exp(-z(i1)))
    activ(i1) = nint(activ(i1))
    if (activ(i1)==y_test(i1)) then
      ncct = ncct + 1
    end if
  end do
  terror = 1.d0 - (dble(ncct)/dble(d))
end subroutine tst

subroutine nnmodel(fvec,n,m,d,c,cgrad)
  implicit none
  integer, intent(in) :: n,m,d !training data and inner layer sizes
  real(kind=8), dimension(m*(n+2)+1), intent(in) :: fvec !fitting parameters
  real(kind=8), intent(out) :: c !cost
  real(kind=8), dimension(m*(n+2)+1), intent(out) :: cgrad !gradient of cost
  integer :: i1,j1,l1
  real(kind=8), dimension(m,n) :: w_inner
  real(kind=8), dimension(m) :: b_inner,w_outer
  real(kind=8) :: dinv,b_outer
  !Declare other variables as needed
  real(kind=8), dimension(m,d) :: z_inner,a_inner,g_inner
  real(kind=8), dimension(d) :: z_outer,a_outer,e_outer,g_outer,eg_outer
  real(kind=8) :: dcdb_outer
  real(kind=8),dimension(m) :: dcdw_outer
  real(kind=8), dimension(m) :: dcdb_inner
  real(kind=8), dimension(m,n) :: dcdw_inner

  dinv = 1.d0/dble(d)

  !unpack fitting parameters (use if needed)
  do i1=1,n
    j1 = (i1-1)*m+1
    w_inner(:,i1) = fvec(j1:j1+m-1) !inner layer weight matrix
  end do
  b_inner = fvec(n*m+1:n*m+m) !inner layer bias vector
  w_outer = fvec(n*m+m+1:n*m+2*m) !output layer weight vector
  b_outer  = fvec(n*m+2*m+1) !output layer bias

  !Add code to compute c and cgrad

  !Compute inner layer activation vector, a_inner
  z_inner = matmul(w_inner,nm_x)
  do i1=1,d
    z_inner(:,i1) = z_inner(:,i1) + b_inner
  end do
  a_inner = 1.d0/(1.d0 + exp(-z_inner))

  !Compute outer layer activation (a_outer) and cost
  z_outer = matmul(w_outer,a_inner) + b_outer
  a_outer = 1.d0/(1.d0+exp(-z_outer))
  e_outer = a_outer-nm_y
  c = 0.5d0*dinv*sum((e_outer)**2)

  !Compute dc/db_outer and dc/dw_outer
  g_outer = a_outer*(1.d0-a_outer)
  eg_outer = e_outer*g_outer
  dcdb_outer = dinv*sum(eg_outer)
  dcdw_outer = dinv*matmul(a_inner,eg_outer)

  !Compute dc/db_inner and dc/dw_inner
  g_inner = a_inner*(1.d0-a_inner)
  dcdb_inner = dinv*w_outer*matmul(g_inner,eg_outer)
  do l1 = 1,n
    dcdw_inner(:,l1) = dinv*w_outer*matmul(g_inner,(nm_x(l1,:)*eg_outer))
  end do

  !Pack gradient into cgrad
  do i1=1,n
    j1 = (i1-1)*m+1
    cgrad(j1:j1+m-1) = dcdw_inner(:,i1)
  end do
  cgrad(n*m+1:n*m+m) = dcdb_inner
  cgrad(n*m+m+1:n*m+2*m) = dcdw_outer
  cgrad(n*m+2*m+1) = dcdb_outer

end subroutine nnmodel

!Compute test error provided fitting parameters
!and with testing data stored in nm_x_test and nm_y_test
subroutine run_nnmodel(fvec,n,m,d,test_error)
  implicit none
  integer, intent(in) :: n,m,d !training data and inner layer sizes
  real(kind=8), dimension(m*(n+2)+1), intent(in) :: fvec !fitting parameters
  real(kind=8), intent(out) :: test_error !test_error
  integer :: i1,j1
  real(kind=8), dimension(m,n) :: w_inner
  real(kind=8), dimension(m) :: b_inner,w_outer
  real(kind=8) :: b_outer
  !Declare other variables as needed
  real(kind=8), dimension(m,d) :: z_inner,a_inner
  real(kind=8), dimension(d) :: z_outer,a_outer
  integer, dimension(d) :: e_outer


  !unpack fitting parameters (use if needed)
  do i1=1,n
    j1 = (i1-1)*m+1
    w_inner(:,i1) = fvec(j1:j1+m-1) !inner layer weight matrix
  end do
  b_inner = fvec(n*m+1:n*m+m) !inner layer bias vector
  w_outer = fvec(n*m+m+1:n*m+2*m) !output layer weight vector
  b_outer  = fvec(n*m+2*m+1) !output layer bias


  !Compute inner layer activation vector, a_inner
  z_inner = matmul(w_inner,nm_x)
  do i1=1,d
    z_inner(:,i1) = z_inner(:,i1) + b_inner
  end do
  a_inner = 1.d0/(1.d0 + exp(-z_inner))

  !Compute outer layer activation (a_outer) and cost
  z_outer = matmul(w_outer,a_inner) + b_outer
  a_outer = 1.d0/(1.d0+exp(-z_outer))
  e_outer = nint(a_outer-nm_y)

  test_error = dble(sum(e_outer))/dble(d)

end subroutine run_nnmodel

subroutine sgd(fvec_guess,n,m,d,alpha,fvec)
  implicit none
  integer, intent(in) :: n,m,d
  real(kind=8), dimension(:), intent(in) :: fvec_guess

  real(kind=8), intent(in) :: alpha
  real(kind=8), dimension(size(fvec_guess)), intent(out) :: fvec
  integer :: i1, j1, i1max=400
  real(kind=8) :: c

  real(kind=8), dimension(size(fvec_guess)) :: cgrad

  real(kind=8), allocatable, dimension(:,:) :: xfull
  integer, allocatable, dimension(:) :: yfull

  real(kind=8), dimension(d+1) :: r

  integer, dimension(d+1) :: j1array

  !store full nm_x,nm_y
  allocate(xfull(size(nm_x,1),size(nm_x,2)),yfull(size(nm_y)))
  xfull = nm_x
  yfull = nm_y

  !will only use one image at a time, so need to reallocate nm_x,nm_y
  call data_init(n,1)

  fvec = fvec_guess
  do i1=1,i1max
    call random_number(r)
    j1array = floor(r*d+1.d0) !d random integers falling between 1 and d (inclusive); will compute c, cgrad for one image at a time cycling through these integers

    do j1 = 1,d
      nm_x(:,1) = xfull(:,j1array(j1))
      nm_y = yfull(j1array(j1))

      !compute cost and gradient with randomly selected image
      if (m==0) then
        call snmodel(fvec,n,1,c,cgrad)
      else
        call nnmodel(fvec,n,m,1,c,cgrad)
      end if
      fvec = fvec - alpha*cgrad !update fitting parameters using gradient descent step
    end do

    if (mod(i1,50)==0) print *, 'completed epoch # ', i1

  end do

 !reset nm_x,nm_y to intial state at beginning of subroutine
  call data_init(size(xfull,1),size(xfull,2))
  nm_x = xfull
  nm_y = yfull
  deallocate(xfull,yfull)
end subroutine sgd


end module nmodel
